libs/pika/async_cuda/include/pika/async_cuda/then_on_host.hpp

```diff
@@ -23,44 +23,59 @@
 namespace pika::cuda::experimental {
     namespace then_on_host_detail {
         template <typename Receiver, typename F>
-        struct then_on_host_receiver
+        struct then_on_host_receiver_impl
+        {
+            struct then_on_host_receiver_type;
+        };
+
+        template <typename Receiver, typename F>
+        using then_on_host_receiver =
+            typename then_on_host_receiver_impl<Receiver,
+                F>::then_on_host_receiver_type;
+
+        template <typename Receiver, typename F>
+        struct then_on_host_receiver_impl<Receiver,
+            F>::then_on_host_receiver_type
         {
             PIKA_NO_UNIQUE_ADDRESS std::decay_t<Receiver> receiver;
             PIKA_NO_UNIQUE_ADDRESS std::decay_t<F> f;
             cuda_scheduler sched;
 
             template <typename Receiver_, typename F_>
-            then_on_host_receiver(
+            then_on_host_receiver_type(
                 Receiver_&& receiver, F_&& f, cuda_scheduler sched)
               : receiver(PIKA_FORWARD(Receiver_, receiver))
               , f(PIKA_FORWARD(F_, f))
               , sched(PIKA_MOVE(sched))
             {
             }
 
-            then_on_host_receiver(then_on_host_receiver&&) = default;
-            then_on_host_receiver& operator=(then_on_host_receiver&&) = default;
-            then_on_host_receiver(then_on_host_receiver const&) = delete;
-            then_on_host_receiver& operator=(
-                then_on_host_receiver const&) = delete;
+            then_on_host_receiver_type(then_on_host_receiver_type&&) = default;
+            then_on_host_receiver_type& operator=(
+                then_on_host_receiver_type&&) = default;
+            then_on_host_receiver_type(
+                then_on_host_receiver_type const&) = delete;
+            then_on_host_receiver_type& operator=(
+                then_on_host_receiver_type const&) = delete;
 
             template <typename Error>
             friend void tag_invoke(pika::execution::experimental::set_error_t,
-                then_on_host_receiver&& r, Error&& error) noexcept
+                then_on_host_receiver_type&& r, Error&& error) noexcept
             {
                 pika::execution::experimental::set_error(
                     PIKA_MOVE(r.receiver), PIKA_FORWARD(Error, error));
             }
 
             friend void tag_invoke(pika::execution::experimental::set_stopped_t,
-                then_on_host_receiver&& r) noexcept
+                then_on_host_receiver_type&& r) noexcept
             {
                 pika::execution::experimental::set_stopped(
                     PIKA_MOVE(r.receiver));
             }
 
             template <typename... Ts>
-            void set_value(Ts&&... ts) noexcept
+            friend void tag_invoke(pika::execution::experimental::set_value_t,
+                then_on_host_receiver_type&& r, Ts&&... ts) noexcept
```

Comment:
Check if this is still a safe change. nvcc previously had problems with the `tag_invoke` overload being here instead of outside the class.

Reply:
This seems to be ok, at least with our current CI configurations. This is somewhat easy to revert if needed, but I'd do so only if we find the need for it.